core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -297,10 +418,16 @@ class LogSegment(val log: FileMessageSet,
   def lastModified = log.file.lastModified
 
   /**
+   * The largest timestamp this segment contains.
+   */
+  def largestTimestamp = if (maxTimestampSoFar >= 0) maxTimestampSoFar else lastModified
```

Comment:
That is a good point. I think it should work. 

There is a slight difference. The current patch essentially takes a snapshot of the timestamp when last message is appended to the log segment. So even if user touched the segments later and changed the modification time, it does not affect the log retention. If we change the behavior to only use lastModified of the file, it could change if user touch the file later. I remember our SRE used to do touch the file for some reasons, but I am not sure if this is an important difference.